### PR TITLE
provide a better hint if a table doesn't exist

### PR DESCRIPTION
**Pull request by bharatnc**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Better hints when a table doesn't exist.

When a table doesn't exist, search if the table exists in other databases and include the database and table in the hint. For example:
```
SELECT * FROM views;

Received exception from server (version 23.11.1):
Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table default.VIEWS does not exist. Maybe you meant INFORMATION_SCHEMA.VIEWS?. (UNKNOWN_TABLE)

DROP TABLE query_log;

Received exception from server (version 23.11.1):
Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table default.query_log does not exist. Maybe you meant system.query_log?. (UNKNOWN_TABLE)
```

Related to: #51168
> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 2c86b251dec91e9f0c6a5e61f5d565d907da94e6 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57342/2c86b251dec91e9f0c6a5e61f5d565d907da94e6/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7028534031">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57342/2c86b251dec91e9f0c6a5e61f5d565d907da94e6/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by bharatnc** [comment]
cc: @alexey-milovidov

File: src/Databases/IDatabase.h [hunk 1]
```diff
@@ -170,7 +170,15 @@ class IDatabase : public std::enable_shared_from_this<IDatabase>
     /// Get the table for work. Return nullptr if there is no table.
     virtual StoragePtr tryGetTable(const String & name, ContextPtr context) const = 0;
 
+    using DatabasePtr = std::shared_ptr<IDatabase>;
+    using Databases = std::map<String, DatabasePtr>;
```
Comment (yakov-olkhovskiy) [inline src/Databases/IDatabase.h:174]: we already have these aliases declared below - I think we can just bring them up

**Comment by yariks5s** [comment]
> The implementation kind of duplicates TableNameHints added in #49801 Should we enhance TableNameHints instead?
> 
> cc: @yariks5s

Yes, I think it is better to add this logic to `TableNameHints` so we will know that everything regarding hints is in the one place

**Comment by bharatnc** [comment]
> > The implementation kind of duplicates TableNameHints added in #49801 Should we enhance TableNameHints instead?
> > cc: @yariks5s
> 
> Yes, it is better to add this logic to `TableNameHints` so we will know that everything regarding hints is in the one place

Thank you for the reviewing. I will make the changes.

File: src/Interpreters/DatabaseCatalog.h [hunk 2]
```diff
@@ -392,6 +369,76 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
...
+    /// the databases that are available in the database catalog.
+    std::pair<String, String> getExtendedHintForTable(const String & table_name) const
+    {
+        /// load all available databases from the DatabaseCatalog instance
+        auto & database_catalog = DatabaseCatalog::instance();
+        auto all_databases = database_catalog.getDatabases();
+
+        /// For returning a pair of hint, which will be the database_name and table_name
+        /// that's closest to the given table_name.
+        std::pair<String, String> result;
```
Comment (yariks5s) [inline src/Interpreters/DatabaseCatalog.h:405]: We don't need this
```suggestion

```

File: src/Interpreters/DatabaseCatalog.h [hunk 3]
```diff
@@ -392,6 +369,76 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
...
+            /// if results are empty continue, else, break and return the first instance of
+            /// table_name and database_name that's found.
+            if (results.empty())
+                continue;
+
+            result = std::make_pair(db_name, results[0]);
+            break;
+        }
+
+        return result;
```
Comment (yariks5s) [inline src/Interpreters/DatabaseCatalog.h:425]:
```suggestion
            if (!results.empty())
                return std::make_pair(db_name, results[0]);
        }
```

File: src/Interpreters/DatabaseCatalog.h [hunk 4]
```diff
@@ -392,6 +369,76 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
...
+        {
+            /// this case should be covered already by getHintForTable
+            if (db_name == database->getDatabaseName())
+                continue;
+
+            TableNameHints hints(db, context);
+            auto results = hints.getHints(table_name);
+
+            /// if results are empty continue, else, break and return the first instance of
+            /// table_name and database_name that's found.
```
Comment (yariks5s) [inline src/Interpreters/DatabaseCatalog.h:417]:
```suggestion
            /// if results are not empty, return the first instance of
            /// table_name and database_name that's found.
```

**Review by yariks5s — APPROVED** [review]

**Pull request merged**